### PR TITLE
[refactor] Add Timestamp Validation

### LINKED ISSUE #25: [refactor] Add Timestamp Validation
## 📝 Description

The `set_asset_price` function accepts any timestamp without validation, allowing timestamps far in the future or stale/replayed timestamps. This enables data manipulation attacks and price history corruption.

**Current problematic code** (`rwa_oracle.rs:314-317`):

```rust
fn set_asset_price(env: &Env, asset_id: Asset, price: i128, timestamp: u64) {
    Self::require_admin(env);
    Self::set_asset_price_internal(env, asset_id, price, timestamp);
    // No timestamp validation at all
}
```

**Why this is critical:**
- Timestamps far in the future could be set, breaking `lastprice()` for dependent contracts
- Old timestamps could be replayed (stale data injection attack)
- Out-of-order timestamps corrupt price history, affecting TWAP calculations
- Contracts relying on fresh prices (lending, liquidations) could receive manipulated data

## ✅ Requirements

- Add new error variants `TimestampInFuture = 8` and `TimestampTooOld = 9` to `error.rs`
- Add constant `MAX_TIMESTAMP_DRIFT_SECONDS = 300` (5 minutes) for network latency tolerance
- Reject timestamps more than 5 minutes in the future relative to ledger time
- Reject timestamps that are <= the last recorded timestamp **for that specific asset** (per-asset validation)
- Each asset should have independent timestamp tracking (updating AAPL shouldn't affect TSLA's timestamps)

## 🎯 Acceptance Criteria

- [ ] `TimestampInFuture` error added with code `8`
- [ ] `TimestampTooOld` error added with code `9`
- [ ] Constant `MAX_TIMESTAMP_DRIFT_SECONDS = 300` defined
- [ ] Timestamps > `ledger_time + 300` are rejected with `TimestampInFuture`
- [ ] Timestamps <= last price timestamp for the same asset are rejected with `TimestampTooOld`
- [ ] Different assets can have independent/unrelated timestamps
- [ ] All 6 required tests implemented and passing
- [ ] All existing tests still pass

## 📁 Expected files to change

- `apps/contracts/stellar-contracts/rwa-oracle/src/error.rs` - Add 2 new errors
- `apps/contracts/stellar-contracts/rwa-oracle/src/rwa_oracle.rs` - Add constant and validation logic
- `apps/contracts/stellar-contracts/rwa-oracle/src/test.rs` - Add 6 new tests

## 🧪 Required Tests

Write tests for the following cases:

| Test Name | Scenario | Expected Result |
|-----------|----------|-----------------|
| `test_future_timestamp_rejected` | Ledger at `1000`, set timestamp `4600` (1 hour ahead) | Panic with `Error(Contract, #8)` |
| `test_timestamp_within_drift_accepted` | Ledger at `1000`, set timestamp `1200` (within 5 min) | Success |
| `test_old_timestamp_rejected` | First price at `t=1000`, try second at `t=999` | Panic with `Error(Contract, #9)` |
| `test_same_timestamp_rejected` | First price at `t=1000`, try second at `t=1000` | Panic with `Error(Contract, #9)` |
| `test_newer_timestamp_accepted` | First price at `t=1000`, second at `t=2000` | Success, `lastprice()` returns second |
| `test_different_assets_independent_timestamps` | AAPL at `t=1000`, then TSLA at `t=500` | Success (TSLA has no prior timestamp) |

**Test Structure Hint:**

```rust
#[test]
fn test_timestamp_within_drift_accepted() {
    // 1. Setup: Create env, register contract, initialize
    // 2. Set ledger timestamp: env.ledger().set_timestamp(1000);
    // 3. Action: set_asset_price with timestamp within drift
    // 4. Assert: lastprice() returns the correct timestamp
}
```

## 🔧 Implementation Hint

```rust
// In error.rs:
TimestampInFuture = 8,
TimestampTooOld = 9,

// In rwa_oracle.rs:
const MAX_TIMESTAMP_DRIFT_SECONDS: u64 = 300;

// In set_asset_price_internal, after price validation:
let current_time = env.ledger().timestamp();

if timestamp > current_time + MAX_TIMESTAMP_DRIFT_SECONDS {
    panic_with_error!(env, Error::TimestampInFuture);
}

if let Some(last_price) = Self::lastprice(env, asset_id.clone()) {
    if timestamp <= last_price.timestamp {
        panic_with_error!(env, Error::TimestampTooOld);
    }
}
```

## ✅ Verification Commands

```bash
cd apps/contracts/stellar-contracts
cargo test --package rwa-oracle
cargo build --package rwa-oracle --target wasm32v1-none-unknown --release
stellar contract build
```

---

Thank you for taking this issue! You are helping us make RWAs consumer friendly on Stellar.

---

### PR DESCRIPTION
**Pull request by gelluisaac**
close #25 

This PR adds strict timestamp validation to set_asset_price to prevent stale, future-dated, or out-of-order price updates. These checks close multiple attack vectors that could corrupt price history, break dependent contracts, or enable price manipulation.

What This PR Does

Introduces explicit timestamp validation rules

Enforces per-asset timestamp ordering (no global coupling)

Adds clear, typed error codes for invalid timestamps

Includes comprehensive test coverage for all edge cases

### CODE REVIEW

**Comment by gelluisaac** [comment]
@aguilar1x  conflicts fixed

**Comment by gelluisaac** [comment]
@aguilar1x  PR is Ready

File: apps/contracts/stellar-contracts/rwa-oracle/src/test.rs [hunk 1]
```diff
@@ -267,6 +338,48 @@ fn test_min_positive_price_accepted() {
...
+    assert_eq!(last_price_tsla.timestamp, timestamp2);
+}
+
+// ========== Issues Test | TODO -> ==========
+    let timestamp: u64 = 1_000_000_000;
+    let min_price: i128 = 1;
+
     oracle.set_asset_price(&asset, &1, &1_000_000_000);
     assert_eq!(oracle.lastprice(&asset).unwrap().price, 1);
 }
```
Comment (aguilar1x) [inline apps/contracts/stellar-contracts/rwa-oracle/src/test.rs]: You're disrupting the code's logic with this; I don't know if you moved it, but it's causing errors for this reason.

File: apps/contracts/stellar-contracts/rwa-oracle/src/test.rs [hunk 2]
```diff
@@ -222,6 +228,23 @@ fn test_error_handling() {
...
+    let asset: Asset = Asset::Other(Symbol::new(&e, "NVDA"));
+
+    set_ledger_timestamp(&e, 1000);
+    let price: i128 = 1;
+    let timestamp: u64 = 4600;
+    oracle.set_asset_price(&asset, &price, &timestamp);
+}
+
+#[test]
+fn test_timestamp_within_drift_accepted() {
```
Comment (aguilar1x) [inline apps/contracts/stellar-contracts/rwa-oracle/src/test.rs]: This test makes no sense; look at the issue requirements to see what it needs.

File: apps/contracts/stellar-contracts/rwa-oracle/src/test.rs [hunk 3]
```diff
@@ -232,7 +255,24 @@ fn test_negative_price_rejected() {
...
+
+    set_ledger_timestamp(&e, 1000);
+    let price: i128 = 123;
+    let timestamp: u64 = 1200;
+    oracle.set_asset_price(&asset, &price, &timestamp);
+
+    let last_price = oracle.lastprice(&asset).unwrap();
+    assert_eq!(last_price.price, price);
+    assert_eq!(last_price.timestamp, timestamp);
+}
```
Comment (aguilar1x) [inline apps/contracts/stellar-contracts/rwa-oracle/src/test.rs]: Don't change this test.

File: apps/contracts/stellar-contracts/rwa-oracle/src/test.rs [hunk 4]
```diff
@@ -222,6 +228,23 @@ fn test_error_handling() {
...
+    let e = Env::default();
+    e.mock_all_auths();
+
+    let oracle = create_rwa_oracle_contract(&e);
+    let asset: Asset = Asset::Other(Symbol::new(&e, "NVDA"));
+
+    set_ledger_timestamp(&e, 1000);
+    let price: i128 = 1;
+    let timestamp: u64 = 4600;
+    oracle.set_asset_price(&asset, &price, &timestamp);
```
Comment (aguilar1x) [inline apps/contracts/stellar-contracts/rwa-oracle/src/test.rs]: There are problems instantiating the oracle object, as it is not finding it, and it is also not recognizing "asset", "price", "timestamp".

**Review by aguilar1x — CHANGES_REQUESTED** [review]

**Comment by gelluisaac** [comment]
@aguilar1x  The earlier errors were caused by unresolved merge-conflict markers in test.rs which broke parsing and made variables appear “undefined”. Conflicts are resolved now; existing tests (including test_negative_price_rejected and min-price test) are restored unchanged. Added the 6 timestamp validation tests per the issue requirements, and all tests are now passing.

**Review by aguilar1x — APPROVED** [review]
LGTM!
